File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/minimum_path_sum/MinimumPathSum.java
package main.java.com.jiang.leetcode100.minimum_path_sum;

import java.util.List;
import java.util.Scanner;
import java.util.ArrayList;

class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        if(m == 0) return 0;
        int n = grid[0].length;
        int[] dp = new int[n];
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(i == 0 || j == 0) {
                    dp[j] = (j > 0 ? dp[j - 1] : dp[j]) + grid[i][j];
                } else {
                    dp[j] = Math.min(dp[j], dp[j - 1]) + grid[i][j];
                }
            }
        }
        return dp[n - 1];
    }
}

public class MinimumPathSum {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> inputs = new ArrayList<>();
        while(!scanner.hasNext("#")) {
            inputs.add(scanner.nextLine());
        }
        scanner.close();

        int[][] grid = new int[inputs.size()][inputs.get(0).split(" ").length];
        for(int i = 0; i < inputs.size(); i++) {
            String[] line = inputs.get(i).split(" ");
            for(int j = 0; j < line.length; j++) {
                grid[i][j] = Integer.valueOf(line[j]);
            }
        }

        System.out.println(new Solution().minPathSum(grid));
    }
}

File: Leetcode100/leetcode100_python/MinimumPathSum.py
from typing import List

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m = len(grid)
        if m == 0:
            return 0
        n = len(grid[0])
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue
                if i == 0:
                    grid[i][j] += grid[i][j - 1]
                elif j == 0:
                    grid[i][j] += grid[i - 1][j]
                else:
                    grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])
        return grid[m - 1][n - 1]

if __name__ == "__main__":
    n = int(input())
    grid = []
    for _ in range(n):
        grid.append([int(num) for num in input().split()])
    solution = Solution()
    print(solution.minPathSum(grid))